experimental/Geometry/Schemes.jl

```diff
@@ -0,0 +1,1244 @@
+  R::T		  	# the ambient polynomial ring to model this affine scheme
+  I::MPolyIdeal{U}	# The ideal in R defining the scheme
+
+  # fields for caching
+  name::String # the name of this scheme for printing
+
+  function Spec(k::S, R::T, I::MPolyIdeal{U} ) where{
+			S <: Ring, T <:MPolyRing , U <: MPolyElem}
+    if k != coefficient_ring(R)
+      error( "Base ring of the affine scheme does not coincide with the base ring of the associated algebra." )
+    end
+    # TODO: Implement further plausibility checks to be performed at runtime.
+    return new{S, T, U}(k, R, I )
+  end
+end
+
+###################################################################################
+# Getter functions
+#
+# No caching is needed in this case, since all these variables need to be assigned 
+# at instantiation
+#
+function base_ring(A::Spec)
+  return A.k
+end
+
+function ambient_ring(A::Spec)
+  return A.R
+end
+
+function defining_ideal(A::Spec)
+  return A.I
+end
+
+##################################################################################
+# Affine scheme that arises as the localization of a Spec at a specific 
+# element 'denom' of the coordinate ring, i.e. a principal open subset. 
+#
+# These are implemented in a recursive fashion meaning there is a parent 
+# which is either a classical Spec or another principal open subset, from 
+# which this instance is derived by inverting an element 'denom'. 
+#
+# Such parent structures form a tree which necessarily has a Spec at its 
+# root. Currently, the implementation only allows localizations at elements 
+# denom from the coordinate ring of this root. 
+#
+mutable struct SpecPrincipalOpen{S,T,U} <: AffineScheme{S,T,U}
+  parent::AffineScheme{S,T,U}
+  denom::U  # element of the "ambient" polynomial ring of the root
+  	    # which has been localized in the parent to get this scheme.
+
+  # Fields for caching. These provide the data as a Spec for this 
+  # affine scheme
+  k::S  # the ground field
+  R::T	# the ambient polynomial ring
+  I::MPolyIdeal{U} # the defining ideal
+  pullbackFromParent::AlgHom
+  pullbackFromRoot::AlgHom
+  u::U # The inverse of denom in R/I (one of the variables)
+  name::String
+  #inclusionInParent::AffSchMorphism
+  #inclusionInRoot::AffSchMorphism
+
+  
+  function SpecPrincipalOpen(parent::Union{Spec{S,T,U},SpecPrincipalOpen{S,T,U}}, denom::U) where {S <: Ring, T<:MPolyRing, U<:MPolyElem}
+    x = new{S,T,U}() 
+    x.parent = parent
+    # TODO: Implement a plausibility check: Does denom belong to the coordinate ring of the root?
+    x.denom = denom
+    return x 
+  end 
+end
+
+###############################################################################
+# Getter functions
+
+function PrincipalSubScheme(parent, denom)
+  return SpecPrincipalOpen(parent, denom)
+end
+
+function localize(parent, denom)
+  return SpecPrincipalOpen(parent, denom)
+end
+
+function parent(S::AffineScheme)
+  return S
```

Comment:
I don't think this is compatible with the meaning of `parent` we employ elsewhere: elements have parents. But e.g. a submodule doesn't have a parent, it has a `supermodule`.

That said, in GAP, we'd also use `parent` in the same way you use it here. I could live with that, but I am not sure if there are strings attached...

So perhaps this should be `superscheme`? Or perhaps we can relax and allow `parent`? Perhaps @fieker @thofma @wbhart have thoughts on this?

Reply:
Good point. And close to the 'identity'-problem encountered earlier. We'll come up with something there. But yes: Suggestions are very welcome.